tools/fp20compiler/main.cpp

```diff
@@ -34,27 +34,80 @@ nvparse_errors errors;
 int line_number;
 char * myin = 0;
 
+static char* find_at_line_start(char* haystack, char* cursor, const char* needle) {
+    cursor = strstr(cursor, needle);
+    while(cursor != NULL) {
+
+        // Accept if candidate is at beginning of file or line
+        if ((cursor == haystack) || (cursor[-1] == '\n')) {
+            return cursor;
+        }
+
+        cursor = strstr(cursor + 1, needle);
+    }
+    return NULL;
+}
+
 void translate(const char* s) {
     char* ns = strdup(s);
 
-    // printf("-- %s\b", s);
+    // Keep a cursor for line-counting
+    char* line_cursor = ns;
+    unsigned int line = 1;
 
-    char* ts_start = strstr(ns, "!!TS1.0");
-    if (ts_start) {
-        char* ts_end = strstr(ts_start, "// End of program");
-        assert(ts_end);
-        *ts_end = 0;
+    // Look for the first shader magic
+    char* shader_start = find_at_line_start(ns, ns, "!!");
 
-        ts10_init(ts_start);
-        ts10_parse();
-
-        *ts_end = '/';
+    // Warn the user if we couldn't find any shader at all
+    if (shader_start == NULL) {
+        fprintf(stderr, "no shaders found\n");
     }
 
-    char* rc_start = strstr(ns, "!!RC1.0");
-    if (rc_start) {
-        rc10_init(rc_start);
-        rc10_parse();
+    // Loop until we can't find a shader anymore
+    while(shader_start != NULL) {
+
+        // The next shader magic marks the end of the current shader; find it
+        char* line_end = strchr(shader_start, '\n');
+        char* next_shader_start = find_at_line_start(ns, line_end + 1, "!!");
+
+        // Create a copy of the current shader magic
+        if (line_end) { *line_end = '\0'; }
```

Comment:
I'm no fan of one-liner `if`s.

Reply:
Me neither, but this is quick back-and-forth patching for `strdup`, so I think it's fine.
Willing to change if someone else has a strong opinion on this.

*I think what's worse, is that this isn't described in a comment, when the same change is documented below, when modifying `*next_shader_start`.*

Reply:
Couldn't this patching be replaced with `strndup`?

Reply:
Yes, sort-of.

I'd still have to differentiate the case where no line-ending exists (it implies end of line is at file ending instead). That makes the code slightly awkward, too.

However, more importantly, the old code already copies all input data for `ns`, to allow patching.
The same style of patching is also done for `next_shader_start` below, which can not be avoided without copying large chunks of input data again.
The patching for `line_end` is thereby consistent with the rest of the code.

However, I also agree that this patching isn't ideal. It might be worth to consider this for a future refactor though (remove `ns` and `strndup` each individual shader / magic to avoid patching).